### PR TITLE
WT-8121 Adding a new cppsuite test that does a large number of insertions over a long period of time.

### PR DESCRIPTION
**Pull request by luke-pearson**
- Changing the thread_context logic so that true means success and false means failure, this is because when a wiredtiger transaction commit fails it rolls back the transaction internally which I didn't anticipate when I wrote the "rollback_required" logic.
- Adding a new test.
- Modifying logic that relies on the thread context logic.

### CODE REVIEW

File: dist/test_data.py [hunk 1]
```diff
@@ -215,6 +215,7 @@ def __ge__(self, other):
 
 methods = {
     'example_test' : Method(test_config),
```
Comment (etienneptl) [inline dist/test_data.py:217]: Alphabetize

File: test/cppsuite/configs/burst_inserts_stress.txt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+),
+runtime_monitor=
+(
+    enabled=false
+),
+timestamp_manager=
+(
+    enabled=true,
+    oldest_lag=10,
+    op_rate=1s,
```
Comment (etienneptl) [inline test/cppsuite/configs/burst_inserts_stress.txt:18]: This is the default value, you dont need it

File: test/cppsuite/configs/burst_inserts_stress.txt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    enabled=true,
+    op_rate=60s
+),
+runtime_monitor=
+(
+    enabled=false
+),
+timestamp_manager=
+(
+    enabled=true,
```
Comment (etienneptl) [inline test/cppsuite/configs/burst_inserts_stress.txt:16]: This is the default value, you dont need it

File: test/cppsuite/configs/burst_inserts_stress.txt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        key_size=200,
+        op_rate=10s,
+        ops_per_transaction=(max=2000,min=1000),
+        thread_count=40,
+        value_size=20000
+    )
+),
+workload_tracking=
+(
+    enabled=false,
```
Comment (etienneptl) [inline test/cppsuite/configs/burst_inserts_stress.txt:42]: Is it causing issue to have it enabled ?

File: test/cppsuite/test_harness/workload/database_operation.cxx [hunk 5]
```diff
@@ -49,16 +49,14 @@ populate_worker(thread_context *tc)
...
+        uint64_t j = 0;
+        while (j < tc->key_count) {
             tc->transaction.begin();
-            if (tc->insert(cursor, coll.id, i)) {
-                /* We failed to insert, rollback our transaction and retry. */
-                tc->transaction.rollback();
-                --i;
-                continue;
+            if (tc->insert(cursor, coll.id, j)) {
+                if (tc->transaction.commit()) {
```
Comment (etienneptl) [inline test/cppsuite/test_harness/workload/database_operation.cxx:56]: You don't need the curly brackets here.

File: test/cppsuite/configs/burst_inserts_stress.txt (hunk 4 above)
Comment (luke-pearson) [reply test/cppsuite/configs/burst_inserts_stress.txt:42]: This test is insert only and generates close to 1TB of data so tracking would just double that.

File: test/cppsuite/test_harness/workload/database_operation.cxx (hunk 5 above)
Comment (luke-pearson) [reply test/cppsuite/test_harness/workload/database_operation.cxx:56]: I thought we decided we were always doing curly brackets cause we're now doing C++ style?

Comment (etienneptl) [reply test/cppsuite/test_harness/workload/database_operation.cxx:56]: True, my bad.

File: test/cppsuite/test_harness/workload/thread_context.cxx [hunk 6]
```diff
@@ -81,22 +82,25 @@ transaction_context::try_begin(const std::string &config)
...
           "Failed to commit transaction in commit, received error code: " + std::to_string(ret));
-        _needs_rollback = true;
-    } else {
-        _op_count = 0;
-        _in_txn = false;
     }
-    return (_needs_rollback);
+    _op_count = 0;
+    _in_txn = false;
+    return (success);
```
Comment (etienneptl) [inline test/cppsuite/test_harness/workload/thread_context.cxx:103]: Instead of having `success`, could we return `ret == 0`?

File: test/cppsuite/test_harness/workload/thread_context.h [hunk 7]
```diff
@@ -136,15 +136,17 @@ class thread_context {
     /*
      * Generic update function, takes a collection_id and key, will generate the value.
      *
-     * Returns true if a rollback is required.
+     * Returns true if the operation was successful, a return value of false implies the transaction
```
Comment (etienneptl) [inline test/cppsuite/test_harness/workload/thread_context.h:139]: Use the imperative tense in description, `Return` *

File: test/cppsuite/test_harness/workload/thread_context.h [hunk 8]
```diff
@@ -136,15 +136,17 @@ class thread_context {
...
+     * needs to be rolled back.
      */
     bool update(scoped_cursor &cursor, uint64_t collection_id, const std::string &key);
 
     /*
      * Generic insert function, takes a collection_id and key_id, will generate the value. If a
      * timestamp is not specified, the timestamp manager will generate one.
      *
-     * Returns true if a rollback is required.
+     * Returns true if the operation was successful, a return value of false implies the transaction
```
Comment (etienneptl) [inline test/cppsuite/test_harness/workload/thread_context.h:148]: Imperative tense here too

File: test/cppsuite/test_harness/workload/thread_context.h (hunk 7 above)
Comment (luke-pearson) [reply test/cppsuite/test_harness/workload/thread_context.h:139]: For all of the instances of returns in this file or just these two?

File: test/cppsuite/tests/burst_inserts.cxx [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            auto burst_start = std::chrono::system_clock::now();
+            while (tc->running() && std::chrono::system_clock::now() - burst_start < std::chrono::seconds(90)) {
+                tc->transaction.try_begin();
+                cc.write_cursor->set_key(cc.write_cursor.get(), tc->key_to_string(start_key + added_count).c_str());
+                cc.write_cursor->search(cc.write_cursor.get());
+
+
+                /* A return value of true implies the insert was successful. */
+                if (!tc->insert(cc.write_cursor, cc.coll.id, start_key + added_count)) {
+                    tc->transaction.rollback();
```
Comment (jiechenbo) [inline test/cppsuite/tests/burst_inserts.cxx:95]: I believe you also need to do `tc->key_to_string(start_key + added_count).c_str() `as well.

File: test/cppsuite/tests/burst_inserts.cxx [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                }
+                added_count++;
+
+                /* Walk our random reader intended to generate cache pressure. */
+                int ret = 0;
+                if ((ret = cc.read_cursor->next(cc.read_cursor.get())) != 0) {
+                    if (ret == WT_NOTFOUND) {
+                        cc.read_cursor->reset(cc.read_cursor.get());
+                    } else if (ret == WT_ROLLBACK) {
+                        tc->transaction.rollback();
```
Comment (jiechenbo) [inline test/cppsuite/tests/burst_inserts.cxx:107]: I wonder if you need to cursor reset() here too? Not sure about the interaction when transaction rolls back.

File: test/cppsuite/configs/burst_inserts_stress.txt [hunk 11]
```diff
@@ -0,0 +1,40 @@
+# Used as a stress test for the framework.
+duration_seconds=14400,
+cache_size_mb=2000,
+compression_enabled=false,
```
Comment (etienneptl) [inline test/cppsuite/configs/burst_inserts_stress.txt:4]: It is false by default

File: test/cppsuite/tests/burst_inserts.cxx [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+ */
+
+#include "test_harness/test.h"
+#include "test_harness/workload/random_generator.h"
+#include "test_harness/timestamp_manager.h"
+
+using namespace test_harness;
+
+/*
+ * Class that defines operations that do nothing as an example. This shows how database operations
```
Comment (jiechenbo) [inline test/cppsuite/tests/burst_inserts.cxx:36]: This needs updating to reflect the test

**Review by jiechenbo — CHANGES_REQUESTED** [review]

File: test/cppsuite/test_harness/workload/database_operation.cxx (hunk 5 above)
Comment (etienneptl) [inline test/cppsuite/test_harness/workload/database_operation.cxx:56]: Could we do the following logic instead of the current one:

- `break` if `insert` fails
- `commit` outside the while loop if `j >= tc->key_count`
- `testutil_assert(commit == true)`

File: test/cppsuite/test_harness/workload/database_operation.cxx [hunk 13]
```diff
@@ -161,27 +159,26 @@ database_operation::insert_operation(thread_context *tc)
         /* Collection cursor. */
         auto &cc = ccv[counter];
         while (tc->transaction.active() && tc->running()) {
-            /* Insert a key value pair. */
-            bool rollback_required = tc->insert(cc.cursor, cc.coll.id, start_key + added_count);
-            if (!rollback_required) {
+            /* Insert a key value pair, rolling back the transaction if required. */
+            if (!tc->insert(cc.cursor, cc.coll.id, start_key + added_count)) {
+                added_count = 0;
+                tc->transaction.rollback();
```
Comment (etienneptl) [inline test/cppsuite/test_harness/workload/database_operation.cxx:165]: I am thinking that if absolutely no insert operations are successful, the test will still be successful since:

- There is nothing inserted, hence nothing to check
- Nothing to read
- Nothing to update

Comment (etienneptl) [reply test/cppsuite/test_harness/workload/database_operation.cxx:165]: It could be another ticket but should we allow a maximum number of failures?

Comment (luke-pearson) [reply test/cppsuite/test_harness/workload/database_operation.cxx:165]: Yeah this doesn't actually test anything per se it just aims to stress WT and see if it creates a failure internally.

File: test/cppsuite/test_harness/workload/thread_context.h (hunk 7 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/workload/thread_context.h:139]: Could be great to fix the whole file since you're on it.

File: test/cppsuite/tests/burst_inserts.cxx [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        std::vector<collection_cursor> ccv;
+        uint64_t collection_count = tc->db.get_collection_count();
+        uint64_t collections_per_thread = collection_count / tc->thread_count;
+        /* Must have unique collections for each thread. */
+        testutil_assert(collection_count % tc->thread_count == 0);
+        for (int i = tc->id * collections_per_thread;
+             i < (tc->id * collections_per_thread) + collections_per_thread && tc->running(); ++i) {
+            collection &coll = tc->db.get_collection(i);
+            ccv.push_back({coll, std::move(tc->session.open_scoped_cursor(coll.name.c_str())),
+              std::move(tc->session.open_scoped_cursor(coll.name.c_str(), "next_random=true"))});
```
Comment (etienneptl) [inline test/cppsuite/tests/burst_inserts.cxx:76]: Can you add a comment about why you decided to enable `next_random` please?

File: test/cppsuite/tests/burst_inserts.cxx [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+        uint64_t counter = 0;
+        while (tc->running()) {
+            uint64_t start_key = ccv[counter].coll.get_key_count();
+            uint64_t added_count = 0;
+            bool committed = true;
+            auto &cc = ccv[counter];
+            auto burst_start = std::chrono::system_clock::now();
+            while (tc->running() &&
+              std::chrono::system_clock::now() - burst_start < std::chrono::seconds(90)) {
```
Comment (etienneptl) [inline test/cppsuite/tests/burst_inserts.cxx:87]: Use a variable instead of the magic number `90`

File: test/cppsuite/tests/burst_inserts.cxx (hunk 9 above)
Comment (etienneptl) [reply test/cppsuite/tests/burst_inserts.cxx:95]: You could create a variable `key_str = tc->key_to_string(start_key + added_count)` and then call `key_str.c_str()`

File: test/cppsuite/tests/burst_inserts.cxx [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+                }
+
+                if (tc->transaction.can_commit()) {
+                    if (tc->transaction.commit()) {
+                        cc.coll.increase_key_count(added_count);
+                        start_key = cc.coll.get_key_count();
+                    }
+                    added_count = 0;
+                }
+                std::this_thread::sleep_for(std::chrono::milliseconds(10));
```
Comment (etienneptl) [inline test/cppsuite/tests/burst_inserts.cxx:122]: Why is this required? Add a comment to explain

File: test/cppsuite/tests/burst_inserts.cxx [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+                    cc.coll.increase_key_count(added_count);
+                    start_key = cc.coll.get_key_count();
+                }
+                added_count = 0;
+            }
+
+            testutil_check(cc.write_cursor->reset(cc.write_cursor.get()));
+            testutil_check(cc.read_cursor->reset(cc.read_cursor.get()));
+            counter++;
+            if (counter == collections_per_thread)
```
Comment (etienneptl) [inline test/cppsuite/tests/burst_inserts.cxx:138]: Defensive code, use `>=` instead of `==`.

File: test/cppsuite/tests/burst_inserts.cxx (hunk 10 above)
Comment (luke-pearson) [reply test/cppsuite/tests/burst_inserts.cxx:107]: No need to do a reset here.

File: test/cppsuite/test_harness/workload/database_operation.cxx (hunk 5 above)
Comment (luke-pearson) [reply test/cppsuite/test_harness/workload/database_operation.cxx:56]: Whats the goal of that logic, there's no guarantee your commit will succeed either?

Comment (etienneptl) [reply test/cppsuite/test_harness/workload/database_operation.cxx:56]: With the current implementation, we could stay blocked in the while loop.

File: test/cppsuite/tests/burst_inserts.cxx (hunk 17 above)
Comment (luke-pearson) [reply test/cppsuite/tests/burst_inserts.cxx:138]: Realistically if counter is ever greater than collections per thread shouldn't we assert as we don't expect that? By doing `>=` you're allowing unexpected behavior.

File: test/cppsuite/tests/burst_inserts.cxx (hunk 16 above)
Comment (luke-pearson) [reply test/cppsuite/tests/burst_inserts.cxx:122]: It was going too quickly but yeah i'll add a comment.

File: test/cppsuite/tests/burst_inserts.cxx (hunk 17 above)
Comment (etienneptl) [reply test/cppsuite/tests/burst_inserts.cxx:138]: You're right, happy with the assertion.

Comment (etienneptl) [reply test/cppsuite/tests/burst_inserts.cxx:138]: Can you fix this one too https://github.com/wiredtiger/wiredtiger/blob/develop/test/cppsuite/test_harness/workload/database_operation.cxx#L191 ?

Comment (luke-pearson) [reply test/cppsuite/tests/burst_inserts.cxx:149]: Sure

File: dist/test_data.py [hunk 18]
```diff
@@ -215,7 +215,9 @@ def __ge__(self, other):
 
 methods = {
     'base_test' : Method(test_config),
-    'burst_inserts' : Method(test_config),
+    'burst_inserts' : Method(test_config + [
+                        Config("burst_duration", 90, r'''
```
Comment (etienneptl) [inline dist/test_data.py:219]: This is a great example of how we can add a specific field to a configuration, nice!

File: test/cppsuite/test_harness/workload/database_operation.cxx [hunk 19]
```diff
@@ -161,27 +161,26 @@ database_operation::insert_operation(thread_context *tc)
         /* Collection cursor. */
         auto &cc = ccv[counter];
         while (tc->transaction.active() && tc->running()) {
-            /* Insert a key value pair. */
-            bool rollback_required = tc->insert(cc.cursor, cc.coll.id, start_key + added_count);
-            if (!rollback_required) {
+            /* Insert a key value pair, rolling back the transaction if required. */
+            if (!tc->insert(cc.cursor, cc.coll.id, start_key + added_count)) {
```
Comment (etienneptl) [inline test/cppsuite/test_harness/workload/database_operation.cxx:165]: Can we check for `true` instead here? Makes more sense to me but it's only my opinion...

Comment (luke-pearson) [reply test/cppsuite/test_harness/workload/database_operation.cxx:165]: Wouldn't we check for false here? I.e. tc->insert == false?

Comment (etienneptl) [reply test/cppsuite/test_harness/workload/database_operation.cxx:165]: Sorry I was not clear. Can we do something like this:
```
if(tc->insert(cc.cursor, cc.coll.id, start_key + added_count)) {
<code_when_insert_is_successful>
} else {
added_count = 0;
tc->transaction.rollback();
}
```

File: test/cppsuite/tests/burst_inserts.cxx (hunk 17 above)
Comment (luke-pearson) [reply test/cppsuite/tests/burst_inserts.cxx:149]: d0dc22c33

File: test/cppsuite/test_harness/workload/database_operation.cxx (hunk 5 above)
Comment (luke-pearson) [reply test/cppsuite/test_harness/workload/database_operation.cxx:56]: I've ran into cases where the commit fails due to cache pressure, I don't think we want to die there if we have some cache pressure. So can we leave it as it for now?

File: test/cppsuite/tests/burst_inserts.cxx [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+            scoped_cursor write_cursor;
+        };
+
+        /* Collection cursor vector. */
+        std::vector<collection_cursor> ccv;
+        uint64_t collection_count = tc->db.get_collection_count();
+        uint64_t collections_per_thread = collection_count / tc->thread_count;
+        /* Must have unique collections for each thread. */
+        testutil_assert(collection_count % tc->thread_count == 0);
+        for (int i = tc->id * collections_per_thread;
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply test/cppsuite/tests/burst_inserts.cxx]: 57bb047fb

File: test/cppsuite/test_harness/workload/database_operation.cxx (hunk 5 above)
Comment (etienneptl) [reply test/cppsuite/test_harness/workload/database_operation.cxx:56]: Ok, we are coming back to the idea of having a variable that sets a maximum number of times we can fail. Let's leave it for now.

**Review by etienneptl — APPROVED** [review]
LGTM, great work!

**Review by jiechenbo — APPROVED** [review]
LGTM!
